### PR TITLE
Electron v9.1.2

### PR DESCRIPTION
**Pull request by rafeca**
⚠️ This PR is based on https://github.com/desktop/desktop/pull/9766 (I can't use it as the base branch of this PR since it's in a fork).

Related issue: #10085 

## Description

Updates the Electron from v8.4.0 to v9.1.0

This will primarly fix a `read: ENOTCONN` error we're seeing more frequently on Electron 7 due to the node version.

## Breaking changes

**Split `shell.openItem(path)` into synchronous and asynchronous methods ([more info](https://github.com/electron/electron/pull/20682)).**

This has been addressed in this PR (commit https://github.com/desktop/desktop/commit/7d4a65e08772463218f8d653ebb0feee645c1695).

Note that this needs to get fixed at the same time as upgrading to v9 since that breaking change had no transition/deprecation plan.

## Major changes

**Chromium got upgraded from v80 to v83.**

* [Release notes for v81](https://developers.google.com/web/updates/2020/04/nic81)
* [Release notes for v83](https://developers.google.com/web/updates/2020/05/nic83)

*(Chromium v82 release got cancelled due to covid-19)*

**Node.js got upgraded from v12.13.0 to v12.4.1.**

* [Release notes for v12.13.1](https://nodejs.org/en/blog/release/v12.13.1/)
* [Release notes for v12.14.0](https://nodejs.org/en/blog/release/v12.14.0/)
* [Release notes for v12.14.1](https://nodejs.org/en/blog/release/v12.14.1/)

**V8 got upgraded from v8.0 to v8.3**.

* [Release notes for v8.1](https://v8.dev/blog/v8-release-81)
* [Release notes for v8.3](https://v8.dev/blog/v8-release-83)

*(V8 v8.2 release got cancelled due to covid-19)*

**Chromium has changed the styling of form controls ([more info](https://developers.google.com/web/updates/2020/05/nic83#forms)).**

I noticed a few subtle differences in the app caused by this (not big changes though since we have custom styling for most of the form controls):

| **Before**        | **After**    |
| ------------- | ----- |
| Screen Shot 2020-07-17 at 13 21 05 | Screen Shot 2020-07-17 at 13 17 54 |
| Screen Shot 2020-07-17 at 13 24 41  | Screen Shot 2020-07-17 at 13 24 49 |
| Screen Shot 2020-07-17 at 13 26 27  | Screen Shot 2020-07-17 at 13 26 34 |

I think the only want that we may want to address is the styling of the "clear" icon on search inputs.

## Full changelog

Changelog of all major/minor Electron releases (these links may break when new Electron versions get published since the pagination in their changelog is not stable):

* [v9.0.0](https://www.electronjs.org/releases/stable?version=9&page=2#release-notes-for-900)
* [v9.1.0](https://www.electronjs.org/releases/stable?version=9#release-notes-for-v910)

### Development extensions not working

Currently the development extensions (React devtools, Chromelens, etc) are broken with Electron v9. 

This is due to a big internal change in how extensions work in electron v9 (see [this issue](https://github.com/electron/electron/issues/19447) and [this PR](https://github.com/electron/electron/pull/21814)) which is not documented in the release notes.

There are a few underlying issues with extensions:

1. Extensions cannot be activated when using `file://` URLs ([issue](https://github.com/electron/electron/issues/24011)).
    
    In Desktop we load the app using a `file://` URL so this is completely blocking us.

2. `electron-devtools-installer` needs to add CRX3 support to be able to install React DevTools.

    ~~There's [a PR](https://github.com/MarshallOfSound/electron-devtools-installer/pull/140) open for this, so we need to wait until it's merged and shipped.~~ **Edit:** The PR has been shipped so this is no longer an issue!

3. There are reports of extensions failing to load when activated just after launching Electron.

    This is a Chromium issue ([link to the issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1085215)). I don't yet know if this is going to affect us, but if it does we'll need to wait for a new Chromium release and a new Electron release that includes it.

### Layout changes related to flexbox (again)

With the new version of Chromium we're experiencing more layout issues. All the issues we've lately experienced when upgrading Electron seem to be related to the gradual transition that Chromium is doing towards the new layout engine [LayoutNG](https://www.chromium.org/blink/layoutng).

In this specific case, when using a `SVG` element as a flexbox item, Chrome will not be able to calculate correctly the width of the SVG element and will set it to its default size (300px):

```html
<div style="display: flex; width: 500px; height: 16px;">
  <div style="flex: 1">Flexbox item</div>
  <svg style="height: 16px; flex-grow: 0; flex-shrink: 0; background-color: red" version="1.1" viewBox="0 0 16 16">
    <path d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48L7.48 8z"></path>
  </svg>
</div>
```

The code above gets rendered as:

Screen Shot 2020-07-17 at 17 59 12

Under normal situations, by defining the height of the SVG the render engine should be able to infer its width using the `viewBox` information but in this case it doesn't and it default to the default 300px size.

**⚠️ Only in experimental mode**

The most interesting part is that this only happens when Electron is configured to tell Chromium to enable its experimental features (which [we do in Desktop](https://github.com/desktop/desktop/blob/dcc22d8dc36cd190e97e07d725e5243e7dd45a11/app/src/main-process/app-window.ts#L45) 😅).

This flag enables the Chromium [experimental Web Platform Features](https://github.com/electron/electron/blob/8bbdc224ac3d7756af930ce946a59ce5ff2533b5/shell/browser/web_contents_preferences.cc#L292-L295). I tried enabling this same flag on my Chrome Web browser (via the `about://flags` interface) but couldn't reproduce the issue in Chrome 84 (I'm using [this codepen](https://codepen.io/rafeca/pen/XWXyJmm)), so it's possible that the issue got fixed and it never reached stable Chrome 🙃 

For now I've worked around this problem by wrapping the SVGs that caused problems under a `<div/>` in https://github.com/desktop/desktop/pull/10220/commits/c092ccf9985e63a8e05271702d16b6f35064c3c5. This is a harmless change and even if it didn't fix this specific bug it wouldn't be a change to worry about imho.

As a next step initiative, I think we should disable the experimental Chrome features in Desktop. They were enabled a long time ago [in this PR](https://github.com/desktop/desktop/pull/1126) to overcome a limitation at that time (the `ResizeObserver` API was only enabled as an experimental feature).

This was done when Desktop used v1.6.3, which used Chromium v56.0.2924.87 ([more info](https://www.electronjs.org/releases/stable?version=1&page=9#1.6.0)), and the `ResizeObserver` API [was enabled in Chrome v64](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) so we shouldn't need the experimental flag at least for that.

I didn't want to disable the flag on this PR since it's possible that we're relying on experimental features somewhere else in the app (or in other rendering quirks that happen only when this flag is enabled) and I prefer this PR to focus only on Electron v9.

(As an additional data point, note that the Electron docs don't recommend enabling experimental features - [source](https://www.electronjs.org/docs/all#8-do-not-enable-experimental-features)).

## Release notes

Notes: No notes

### CODE REVIEW

**Comment by rafeca** [comment]
The current build failure is caused by the [`fs-admin`](https://github.com/atom/fs-admin) package, which latest published release doesn't include prebuilt binaries for Electron v9.

Good thing is that there's an already merged [PR](https://github.com/atom/fs-admin/pull/61) that adds support for it, so we only need a new version of the package to get published (I've talked to @shiftkey , the maintainer of the package, and he's planning to publish a new version in the upcoming days).

Once this gets fixed, builds will hopefully start passing 😃

**Comment by shiftkey** [comment]
@rafeca please update `fs-admin` to [`v0.15.0`](https://github.com/atom/fs-admin/releases/tag/v0.15.0) as that includes prebuild support for Electron 9

**Comment by rafeca** [comment]
Seems like CI is passing now! The only blocking thing that we need to solve now is the installation of devtools extensions for development.

**Comment by say25** [comment]
Electron 9.1.1 + 9.1.2 have dropped recently, with bug fixes.

**Comment by rafeca** [comment]
As discussed with @niik, I've ended up disabling the Web Experimental features in this PR (see 
3d84bc8) since we already need to do full regression tests here. This allows us to revert the commit that wrapped SVGs around `divs` (c092ccf9985e63a8e05271702d16b6f35064c3c5).

File: app/src/lib/fix-emoji-spacing.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  '--font-size-lg',
+  '--font-size-xl',
+  '--font-size-xxl',
+  '--font-size-xs',
+]
+
+for (const fontSize of fontSizes) {
+  const span = document.createElement('span')
+  span.setAttribute('style', `font-size: var(${fontSize});`)
+  span.innerHTML = '🤦🏿‍♀️'
```
Comment (niik) [inline app/src/lib/fix-emoji-spacing.ts:28]: innerHTML scares the heck out of me

```suggestion
  span.textContent = '🤦🏿‍♀️'
```

File: app/src/lib/fix-emoji-spacing.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+// a non system-default font to workaround an Chrome
+// issue that causes unexpected spacing on emojis.
+// More info:
+// https://bugs.chromium.org/p/chromium/issues/detail?id=1113293
+
+const container = document.createElement('div')
+container.setAttribute(
+  'style',
+  'visibility: hidden; font-family: Arial !important;'
+)
```
[the remark this answers is not in the document]
Comment (rafeca) [reply app/src/lib/fix-emoji-spacing.ts:11]: Good catch! You're totally right!

File: app/src/lib/fix-emoji-spacing.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  span.setAttribute('style', `font-size: var(${fontSize});`)
+  span.innerHTML = '🤦🏿‍♀️'
+  container.appendChild(span)
+}
+
+document.body.appendChild(container)
+
+// Read the dimensions of the element to force the browser to do a layout.
+// eslint-disable-next-line @typescript-eslint/no-unused-expressions
+container.offsetHeight
```
[the remark this answers is not in the document]
Comment (rafeca) [reply app/src/lib/fix-emoji-spacing.ts:35]: Sounds good to me, and now that I think of it I need to ensure that this statement does not get removed by the minifier in production (it's possible that the `toString()` call also helps in that regard)

[the remark this answers is not in the document]
Comment (rafeca) [reply app/src/lib/fix-emoji-spacing.ts]: I've checked and the statement does not get stripped in production :tada:

**Comment by tierninho** [comment]
Performed a sweep of tests on both OSX/Win and nothing was spotted outside of the two bugs filed (#10340 and #10341). Additionally, I did not directly experience any of the open Electron issue above. All appears to be functioning as intended, so I believe we are ready for the next step.

**Pull request merged**